File: app/assets/javascripts/pageflow/editor/api/file_importers.js
pageflow.FileImporters = pageflow.Object.extend({
  initialize: function() {
    this.importers = {};
  },
  register: function(name, config) {
    if (this._setup) {
      throw 'File importers setup is already finished. Register file importers before setup is finished';
    }

    this.importers[name] = config;
    config.key = name;
  },
  setup: function(serverSideConfigs) {
    this._setup = true;
  },
  find: function(name) {
    if (!this.importers[name]) {
      throw('Could not find file importer with name "' + name +'"');
    }
    return this.importers[name];
  },
  keys: function () {
    return Object.keys(this.importers);
  },
  values: function () {
    return Object.values(this.importers);
  }

});

File: lib/pageflow/file_importers.rb
module Pageflow
  # List of file importers will be maintained using this class
  # File importer plugin is required to register itself to this list
  class FileImporters
    include Enumerable

    def initialize
      clear
    end

    def register(file_importer)
      @file_importers[file_importer.name] = file_importer
    end

    def clear
      @file_importers = {}
    end

    def each(&)
      @file_importers.values.each(&)
    end

    def find_by_name!(name)
      fetch_by_name(name) do
        raise "Unknown file importer with name '#{name}'."
      end
    end

    def fetch_by_name(name, &)
      @file_importers.fetch(name, &)
    end
  end
end
